File: notification-service/app/tracing.py
"""Distributed tracing setup (ADR-0024's 2026-08-02 Update): exports spans
over OTLP/HTTP to the Jaeger container declared in docker-compose.yml.
Additive to app/correlation.py's existing CorrelationId propagation, not a
replacement -- see the ADR update for why. Mirrors api-gateway's/
inventory-service's own setup_tracing, duplicated rather than shared,
since this project keeps no cross-language shared library (only
correlation-commons/-go, and only for the byte-for-byte-identical
CorrelationId infra -- see docs/architecture/observability.md)."""

import logging
import os

from opentelemetry import propagate, trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import SERVICE_NAME, Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

logger = logging.getLogger(__name__)


def setup_tracing(service_name: str) -> None:
    endpoint = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT")
    if not endpoint:
        # No collector configured (e.g. running outside Compose) -- leave
        # the global no-op tracer in place rather than fail startup.
        return

    resource = Resource.create({SERVICE_NAME: service_name})
    provider = TracerProvider(resource=resource)
    exporter = OTLPSpanExporter(endpoint=f"{endpoint}/v1/traces")
    provider.add_span_processor(BatchSpanProcessor(exporter))
    trace.set_tracer_provider(provider)

    logger.info("otel: tracing enabled, exporting to %s", endpoint)


def extract_trace_context(headers: dict | None):
    """Reads an incoming delivery's headers for a traceparent the publisher
    injected -- the RabbitMQ-hop equivalent of _scope_from_properties'
    CorrelationId reuse. Returns a Context usable with
    trace.use_span/start_as_current_span(context=...)."""
    return propagate.extract(headers or {})


tracer = trace.get_tracer("notification-service.messaging")

File: inventory-service/tracing.go
package main

import (
	"context"
	"log"
	"os"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
)

// setupTracing wires the OpenTelemetry SDK to export spans over OTLP/HTTP
// to the Jaeger container declared in docker-compose.yml (see ADR-0024's
// 2026-08-02 Update). Identical shape to api-gateway's own setupTracing --
// duplicated rather than shared, since this is genuinely per-service
// wiring, not the byte-for-byte-identical CorrelationId infra that earns
// correlation-commons-go's exception to "no shared library between
// services" (see docs/architecture/observability.md).
func setupTracing(ctx context.Context, serviceName string) func(context.Context) error {
	endpoint := os.Getenv("OTEL_EXPORTER_OTLP_ENDPOINT")
	if endpoint == "" {
		return func(context.Context) error { return nil }
	}

	exporter, err := otlptracehttp.New(ctx, otlptracehttp.WithEndpointURL(endpoint))
	if err != nil {
		log.Printf("otel: failed to create OTLP exporter, tracing disabled: %v", err)
		return func(context.Context) error { return nil }
	}

	res, err := resource.New(ctx,
		resource.WithAttributes(semconv.ServiceName(serviceName)),
	)
	if err != nil {
		log.Printf("otel: failed to build resource, using default: %v", err)
		res = resource.Default()
	}

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exporter),
		sdktrace.WithResource(res),
	)
	otel.SetTracerProvider(tp)
	otel.SetTextMapPropagator(propagation.TraceContext{})

	log.Printf("otel: tracing enabled, exporting to %s", endpoint)

	return func(shutdownCtx context.Context) error {
		ctx, cancel := context.WithTimeout(shutdownCtx, 5*time.Second)
		defer cancel()
		return tp.Shutdown(ctx)
	}
}
